_release/bat/base_bat/base_bat_test.go

```diff
@@ -311,6 +327,179 @@ func TestDeleteAllInstances(t *testing.T) {
 	}
 }
 
+//Launch an instance created from an image with a customized cloud init
+//
+//Download cirros image, add it to the image service using a custom cloudinit file
+//create a workload, launch an instance of this workload and connect via SSH
+//using the custom user and password
+//
+//instance should accept connection with specified user and password and
+//reply with the user name
+func TestLaunchCustomInstance(t *testing.T) {
+	//creating a new cirros image
+	const name = "cirros-image"
+	ctx, cancelFunc := context.WithTimeout(context.Background(), standardTimeout)
+	defer cancelFunc()
+
+	// TODO:  The only options currently supported by the image service are
+	// ID and Name.  This code needs to be updated when the image service's
+	// support for meta data improves.
+	options := bat.ImageOptions{
+		Name: name,
+	}
+	//downaload cirros image
+	ImageFile, err := os.Create("cirros-0.3.4-x86_64-disk.img")
+	if err != nil {
+		t.Fatalf("Error while creating image file %v", err)
+	}
+	defer ImageFile.Close()
+
+	resp, err := http.Get("http://download.cirros-cloud.net/0.3.4/cirros-0.3.4-x86_64-disk.img")
+	if err != nil {
+		t.Fatalf("Error while downloading image file %v", err)
+	}
+	defer resp.Body.Close()
+	_, err = io.Copy(ImageFile, resp.Body)
+
+	if err != nil {
+		t.Fatalf("Error while saving image file %v", err)
+	}
+
+	img, err := bat.AddImage(ctx, "", "cirros-0.3.4-x86_64-disk.img", &options)
+	if err != nil {
+		t.Fatalf("Unable to add image %v", err)
+	}
+
+	defer deleteCustomData(ctx, img.ID)
+
+	if img.ID == "" || img.Name != name || img.Status != "active" ||
+		img.Visibility != "public" || img.Protected {
+		t.Errorf("Meta data of added image is incorrect")
+	}
+	//get data to create workload yaml file
+	defaults := bat.DefaultResources{
+		VCPUs: 2,
+		MemMB: 128,
+	}
+
+	source := bat.Source{
+		Type: "image",
+		ID:   img.ID,
+	}
+	disk := bat.Disk{
+		Bootable:  true,
+		Source:    &source,
+		Ephemeral: true,
+	}
+
+	WLopts := bat.WorkloadOptions{
+		Description:     "BAT VM Test",
+		VMType:          "qemu",
+		FWType:          "legacy",
+		Defaults:        defaults,
+		CloudConfigFile: "CirrosInit.yaml",
+		Disks:           []bat.Disk{disk},
+	}
+	y, err := yaml.Marshal(WLopts)
+	if err != nil {
+		t.Fatal("fail to read yaml file $v", err)
+	}
+	ioutil.WriteFile("CirrosWorkload.yaml", y, 0644)
```

Comment:
Actually, there's already a function that does all of this work (createWorkloads).  Currently, it's in workloads.go,  so  you'd need to move it to the bat package first.